tpl/lang: Add NumFmt function

moorereason opened the pull request:
numFormat formats a number with a given precision using the requested decimal,
grouping, and negative characters.

Fixes #1444

moorereason commented:
Replaces #1450.

This should be simpler than the previous attempt.  Formats the number using `strconv` and then inserts/replaces  characters where needed.  Should be easier to maintain.

bep commented:
@moorereason if you rebase this, I will merge it. I see many people ask for it, and once we get some aliasing and namespaces in order, I'm not too worried about having "many" funcs. But double check the arg order (and example) so it fits nicely into a pipe.

moorereason commented:
Added a pipe example and rebased.

digitalcraftsman commented on docs/content/templates/functions.md:
```diff
@@ -460,6 +460,19 @@ e.g.
+{{ numFormat "," "." "-" 2 12345.6789 }} → 12.345,68
+{{ numFormat "." "" "-" 6 -12345.6789 }} → -12345.678900
+{{ numFormat "." "," "-" 0 -12345.6789 }} → -12,346
+{{ -12345.6789 | numFormat "." "," "-" 2 }} → -12,346.68
```
Shouldn't the result be `-12,345.68`

bep replied:
While you're in that file, you may also specify what rounding rule used.

moorereason replied:
Doh! typo

moorereason commented:
Pushed a new commit.  Please review.

moorereason commented:
@bep, we can merge this after #3415, but I made some changes from my original PR:

- Moved this function to the `lang` namespace since this is really a localization concern.
- Removed shorthand `numFormat` from funcMap.  Can only access via namespace.
- Renamed from `numFormat` to `NumFmt`.

moorereason commented:
@bep, this is ready for review.

bep commented on docs/content/templates/functions.md:
````diff
@@ -460,6 +460,22 @@ e.g.
+So, with precision set to 0, 1.5 becomes `2`, and 1.4 becomes `1`.
+
+```
+{{ lang.NumFmt "," "." "-" 2 12345.6789 }} → 12.345,68
````
Thinking about this, the signature looks kind of klunky for a template func.

* I kind of want to put the format into my site config, so people can set their own.
* I doubt this will work: `myNumFormat = [",", ".", "-"]`
* Suspect this will be better and then use `strings.Fields`:  `myNumFormat = ", . _"`

Other than that it looks fine.

moorereason replied:
Good thoughts.  How about we build it with the intention that it will work with l10n later.  Like i18n, the options are defined in the site config.

```go
func NumFmt(precision int, n float64, opts ...interface{}) string
```

```
// Assume that they've defined locale number formatting in the site config.
{{ lang.NumFmt 2 123.456 }}
{{ 123.456 | lang.NumFmt 2 }}

// Given multiple locales defined in config, allow to specify which one to use.
{{ lang.NumFmt 2 123.456 "loc:en_US" }}

// Allow use without a locale definition?
{{ lang.NumFmt 2 123.456 "dec:." "grp:," "neg:-" }}
```

We don't have to support all of those options, but the func signature would allow us flexibility in the implementation as we move toward l10n.

If we need to think through l10n before merging this, I'm okay with that.

bep replied:
We have no concept of locale or "locale formatting". What you are saying is that we should wait until we see how that looks like before we revisit this one.

bep replied:
But I think we should just drop the locale debate, and just pack the three arguments into one string. It should be possible to do some magic locale stuff later if needed.

moorereason commented:
@bep, how's this look?  I still need to add site config support.

bep commented:
This looks great. Don't bother with the site config support; we will have to revisit this at some point. With this, people can easily put that config in site params.
